### PR TITLE
[Drugs] Meth Levelling

### LINKED ISSUE #1526: [Feature] Meth Leveling
XP Based Meth Levelling System
Perks: Less Explosion Chance, Faster Producing, More Efficiency
Levels 1-3 planned

### PR DESCRIPTION
**Pull request by maxgestic**
Added Meth Levelling
- Ranks 1-5 with decrease in blow up chance and faster processing and packaging speeds

Closes #1526

### CODE REVIEW

File: resources/usa_meth/cl_meth.lua [hunk 1]
```diff
@@ -198,6 +204,18 @@ AddEventHandler("methJob:getSupplies", function(supplyType)
...
 end)
 
+RegisterNetEvent("methJob:setRank")
+AddEventHandler("methJob:setRank", function(rank)
+    if methRank == nil then
+        methRank = rank
+    else
+        if rank ~= methRank then
+            methRank = rank
+            TriggerEvent("usa:notify", "Level Up! Your meth rank is now "..methRank)
```
Comment (minipunch) [inline resources/usa_meth/cl_meth.lua:214]: Doesn't this mean this notification will be sent every single time the person loads a character with the way you have it now

**Review by minipunch — CHANGES_REQUESTED** [review]

File: resources/usa_meth/cl_meth.lua (hunk 1 above)
Comment (maxgestic) [reply resources/usa_meth/cl_meth.lua:214]: No thats why it checks if the client has their methRank set to nil so when they first load their char it should not send a notification

File: resources/usa_meth/sv_meth.lua [hunk 2]
```diff
@@ -184,3 +189,68 @@ AddEventHandler("methJob:checkUserMoney", function(supplyType)
...
+        db.getDocumentById("methjob", char.get("_id"), function(doc)
+        	local newxp = doc.xp + xp
+        	db.updateDocument("methjob", char.get("_id"), {xp = newxp}, function() end)
+        	setRank(char,newxp)
+        	TriggerClientEvent("usa:notify", char.get("source"), "You have gained "..xp.." Meth XP")
+        end)
+    end)
+end
+
+TriggerEvent('es:addCommand', 'addxp', function(source, args, char, location)
```
[the remark this answers is not in the document]
Comment (maxgestic) [reply resources/usa_meth/sv_meth.lua:248]: no lol

File: resources/usa_meth/sv_meth.lua [hunk 3]
```diff
@@ -56,7 +59,9 @@ AddEventHandler("methJob:methProcessed", function(itemName, securityToken)
 	if char.canHoldItem(methProduced) then
 		char.giveItem(methProduced)
 		TriggerClientEvent("usa:notify", src, "You have successfully processed meth rock into packaged product!")
+		addXP(char,50)
 	else
+		TriggerEvent("interaction:addDroppedItem", methProduced)
```
[the remark this answers is not in the document]
Comment (maxgestic) [reply resources/usa_meth/sv_meth.lua:64]: Yea i dont think it gives it in the client file ill test it to double check but i think it was just missing cos above in "methJob:methProduced" it has the addDroppedItem if the char cant hold the item so i just added it to this one as well

File: resources/usa_meth/sv_meth.lua [hunk 4]
```diff
@@ -184,3 +189,68 @@ AddEventHandler("methJob:checkUserMoney", function(supplyType)
...
+		rank = 4
+	elseif xp >= 4000 then
+		rank = 5
+	end
+	TriggerClientEvent("methJob:setRank", char.get("source"), rank)
+end
+
+function addXP(char, xp)
+	TriggerEvent('es:exposeDBFunctions', function(db)
+        db.getDocumentById("methjob", char.get("_id"), function(doc)
```
[the remark this answers is not in the document]
Comment (maxgestic) [reply resources/usa_meth/sv_meth.lua:239]: Done

File: resources/usa_meth/sv_meth.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxgestic) [reply resources/usa_meth/sv_meth.lua:74]: Tested it and it works good

**Review by minipunch — APPROVED** [review]
I think we might want to increase the XP required to rank up since 1000 xp is only like 10 crafts If I am not mistaken

Might wanna make it like 20 - 30 at least so people don't climb the ranks too easily

But for now I'll merge this as is and you can do that later

**Pull request merged**